Added scenario for Integrating Azure Arc enabled Kubernetes environment with Azure Key Vault (#851)
microsoft/azure_arc: docs/azure_arc_jumpstart/azure_arc_k8s/day2/cluster_api/cluster_api_keyvault_extension/_index.md

    ![Screenshot showing Azure Portal with Azure Arc-enabled Kubernetes resource extensions](./02.png)

* Edit the environment variables in the script to match your environment parameters followed by running the ```. ./azure_keyvault_k8s_extension.sh``` command.

    ![Screenshot parameter examples](./03.png)

    > **Note: The extra dot is due to the shell script has an *export* function and needs to have the vars exported in the same shell session as the rest of the commands.**

   The script will:

  * Login to your Azure subscription using the SPN credentials
  * Add or Update your local _connectedk8s_ and _k8s-extension_ Azure CLI extensions
  * Create Resource Group, Key Vault and sample secret
  * Create Azure Key Vault k8s extension instance
  * Create Kubernetes Secret for supporting SPN Authentication
  * Create Secret Provider Class to read the secrets
  * Deploy the app

* You can now see that Azure Key Vault extensions are now enabled in the extension tab section of the Azure Arc-enabled Kubernetes cluster resource in Azure.

    ![Screenshot extension deployment extensions tab](./04.png)

> * You can also verify the deployment by running the below _kubectl_ commands and see the deployed artifacts in the _kube-system_ namespace.

Reviewer: Please add the commands to run as text values as well using

```shell
```

Author: Added the commands

    ![Screenshot extension and app deployment](./05.png)

## Validate the Key Vault

* To verify that Azure Key Vault Secret Provider is working properly and fetching the secrets, run the below command to show secrets held in secrets-store:

>     ```bash

Reviewer: ```suggestion
    ```shell
```

    kubectl -n hello-arc exec busybox-secrets-sync -- ls /mnt/secrets-store/
    ```

    ![Screenshot Kubernetes App Pod secret store](./06.png)

* Run the below command to print a test secret held in secrets-store:

>     ```bash

Reviewer: ```suggestion
    ```shell
```

    kubectl -n hello-arc exec busybox-secrets-sync -- sh -c 'echo $SECRET_USERNAME'
    ```

    ![Screenshot Kubernetes App Pod secret](./07.png)

## Delete extension instance

* The following command only deletes the extension instance.

>     ```bash

Reviewer: ```suggestion
    ```shell
```

---
type: docs
title: "Integrate Azure Key Vault with Cluster API as an Azure Arc Connected Cluster using Kubernetes extensions"
linkTitle: "Integrate Azure Key Vault with Cluster API as an Azure Arc Connected Cluster using Kubernetes extensions"
weight: 3
description: >
---

## Integrate Azure Key Vault with Cluster API as an Azure Arc Connected Cluster using Kubernetes extensions

The following README will guide you on how to enable [Azure Key Vault](https://docs.microsoft.com/en-us/azure/key-vault/general/overview) for a Cluster API that is projected as an Azure Arc connected cluster.

> In this guide, you will hook the Cluster API to Azure Key Vault by deploying the [Azure Key Vault extension](https://docs.microsoft.com/en-us/azure/aks/csi-secrets-store-driver) and a sample app on your Kubernetes cluster in order to integrate Azure Key Vault as a secrets store with a Kubernetes cluster via a CSI volume. This will also include deploying a Key Vault and sample secret on your Azure subscription.

Reviewer 2: ```suggestion
In this guide, you will hook the Cluster API to Azure Key Vault by deploying the [Azure Key Vault extension](https://docs.microsoft.com/en-us/azure/aks/csi-secrets-store-driver) and a sample app on your Kubernetes cluster in order to integrate Azure Key Vault as a secrets store with a Kubernetes cluster via a [Container Storage Interface (CSI)](https://kubernetes-csi.github.io/docs/) volume. This will also include deploying a Key Vault and sample secret on your Azure subscription.
```

Reviewer 3: Should we update _Azure Key Vault extension_ link to either extension public documentation or upstream docs? This link is referring AKS add-on and might be confusing to user.

> > **Note: This guide assumes you already deployed a Cluster API and connected it to Azure Arc. If you haven't, this repository offers you a way to do so in an automated fashion using a [Shell script](https://azurearcjumpstart.io/azure_arc_jumpstart/azure_arc_k8s/cluster_api/capi_azure/).**

Reviewer 2: Should we use Kubernetes cluster deployed using Cluster API in the doc references here instead of `deployed a Cluster API`? Cluster API is a project for deploying Kubernetes cluster using declarative APIs.

Same comment for other references in the doc

Reviewer 4: @aramase yes. This is a cross-reference to the Jumpstart scenario we support and align with the upstream team

Kubernetes extensions are add-ons for Kubernetes clusters. The extensions feature on Azure Arc-enabled Kubernetes clusters enables usage of Azure Resource Manager based APIs, CLI and portal UX for deployment of extension components (Helm charts in initial release) and will also provide lifecycle management capabilities such as auto/manual extension version upgrades for the extensions.

## Prerequisites

* Clone the Azure Arc Jumpstart repository

    ```shell

>     git clone https://github.com/microsoft/azure_arc.git

Reviewer 5: is this used for the shell script only? should we have users download the script only instead of cloning the entire repo?

Reviewer: This is a standard instruction for all jumpstart scenarios. I would prefer to keep this rather than downloading individual scripts. @Likamrat for any other input.

Reviewer 4: +1 @dkirby-ms

    ```shell
    az login
    az ad sp create-for-rbac -n "<Unique SP Name>" --role contributor
    ```

    For example:

    ```shell
    az ad sp create-for-rbac -n "http://AzureArcK8s" --role contributor
    ```

    Output should look like this:

    ```json
    {
    "appId": "XXXXXXXXXXXXXXXXXXXXXXXXXXXX",
    "displayName": "AzureArcK8s",
    "name": "http://AzureArcK8s",
    "password": "XXXXXXXXXXXXXXXXXXXXXXXXXXXX",
    "tenant": "XXXXXXXXXXXXXXXXXXXXXXXXXXXX"
    }
    ```

    > **Note: The Jumpstart scenarios are designed with as much ease of use in-mind and adhering to security-related best practices whenever possible. It is optional but highly recommended to scope the service principal to a specific [Azure subscription and resource group](https://docs.microsoft.com/en-us/cli/azure/ad/sp?view=azure-cli-latest) as well considering using a [less privileged service principal account](https://docs.microsoft.com/en-us/azure/role-based-access-control/best-practices)**

## Automation Flow

For you to get familiar with the automation and deployment flow, below is an explanation.

> * User has deployed Kubernetes using Cluster API and has it connected as Azure Arc-enabled Kubernetes cluster.

Reviewer 2: ```suggestion
* User has deployed Kubernetes cluster using Cluster API and has it connected as Azure Arc-enabled Kubernetes cluster.
```

> * User is editing the environment variables on the Shell script file (1-time edit) which then be used throughout the extension deployment.

Reviewer 2: ```suggestion
* User is editing the environment variables in the Shell script file (1-time edit) which then be used throughout the extension deployment.
```

* User is running the shell script. The script will use the extension management feature of Azure Arc to deploy the Azure Key Vault Secrets Provider extension on the Azure Arc connected cluster.

> * The script will also deploy the sample app to Azure Arc-enabled Kubernetes cluster

Reviewer 2: Should we also mention the script creates the keyvault and secret?

Author: Added comment about KV and sample secret

    ![Screenshot parameter examples](./03.png)

    > **Note: The extra dot is due to the shell script has an *export* function and needs to have the vars exported in the same shell session as the rest of the commands.**

   The script will:

  * Login to your Azure subscription using the SPN credentials
  * Add or Update your local _connectedk8s_ and _k8s-extension_ Azure CLI extensions
  * Create Resource Group, Key Vault and sample secret
  * Create Azure Key Vault k8s extension instance
  * Create Kubernetes Secret for supporting SPN Authentication
  * Create Secret Provider Class to read the secrets
  * Deploy the app

* You can now see that Azure Key Vault extensions are now enabled in the extension tab section of the Azure Arc-enabled Kubernetes cluster resource in Azure.

    ![Screenshot extension deployment extensions tab](./04.png)

* You can also verify the deployment by running the below _kubectl_ commands and see the deployed artifacts in the _kube-system_ namespace.

    ```shell
    kubectl get pods,secret -n hello-arc
    kubectl get secretproviderclass -n hello-arc

>     kubectl get customresourcedefinitions

Reviewer 3: ```suggestion
    kubectl get customresourcedefinitions secretproviderclasses.secrets-store.csi.x-k8s.io secretproviderclasspodstatuses.secrets-store.csi.x-k8s.io
```

Alternately we can highlight these extension specific CRDs in screenshot.

    ```

    ![Screenshot extension and app deployment](./05.png)

> ## Validate the Key Vault

Reviewer 2: ```suggestion
## Validate the secrets
```

* To verify that Azure Key Vault Secret Provider is working properly and fetching the secrets, run the below command to show secrets held in secrets-store:

    ```shell
    kubectl -n hello-arc exec busybox-secrets-sync -- ls /mnt/secrets-store/
    ```

    ![Screenshot Kubernetes App Pod secret store](./06.png)

* Run the below command to print a test secret held in secrets-store:

    ```shell
    kubectl -n hello-arc exec busybox-secrets-sync -- sh -c 'echo $SECRET_USERNAME'
    ```

    ![Screenshot Kubernetes App Pod secret](./07.png)

## Delete extension instance

* The following command only deletes the extension instance.

    ```shell

>     az k8s-extension delete --name Microsoft.AzureKeyVaultSecretsProvider --cluster-type connectedClusters --cluster-name <cluster-name> --resource-group <resource-group>

Reviewer 3: ```suggestion
    az k8s-extension delete --name <extension-name> --cluster-type connectedClusters --cluster-name <cluster-name> --resource-group <resource-group> -y
```

    ```

> * You can also delete the extensions from the Azure Portal under the extensions section of Azure Arc-enabled Kubernetes cluster resource.

Reviewer 3: ```suggestion
* You can also delete the extension from the Azure Portal under the extensions section of Azure Arc-enabled Kubernetes cluster resource.
```

linkTitle: "Integrate Azure Key Vault with Cluster API as an Azure Arc Connected Cluster using Kubernetes extensions"
weight: 3
description: >
---

## Integrate Azure Key Vault with Cluster API as an Azure Arc Connected Cluster using Kubernetes extensions

The following README will guide you on how to enable [Azure Key Vault](https://docs.microsoft.com/en-us/azure/key-vault/general/overview) for a Cluster API that is projected as an Azure Arc connected cluster.

In this guide, you will hook the Cluster API to Azure Key Vault by deploying the Azure Key Vault extension and a sample app on your Kubernetes cluster in order to integrate Azure Key Vault as a secrets store with a Kubernetes cluster via a [Container Storage Interface (CSI)](https://kubernetes-csi.github.io/docs/) volume. This will also include deploying a Key Vault and sample secret on your Azure subscription.

> **Note: This guide assumes you already deployed a Cluster API and connected it to Azure Arc. If you haven't, this repository offers you a way to do so in an automated fashion using a [Shell script](https://azurearcjumpstart.io/azure_arc_jumpstart/azure_arc_k8s/cluster_api/capi_azure/).**

> Kubernetes extensions are add-ons for Kubernetes clusters. The extensions feature on Azure Arc-enabled Kubernetes clusters enables usage of Azure Resource Manager based APIs, CLI and portal UX for deployment of extension components (Helm charts in initial release) and will also provide lifecycle management capabilities such as auto/manual extension version upgrades for the extensions.

Reviewer 4: ```suggestion
Kubernetes extensions are add-ons for Kubernetes clusters. The extensions feature on Azure Arc-enabled Kubernetes clusters enables usage of Azure Resource Manager based APIs, CLI, and portal UX for the deployment of extension components (Helm charts in initial release) and will also provide lifecycle management capabilities such as auto/manual extension version upgrades for the extensions.
```

## Prerequisites

> * Clone the Azure Arc Jumpstart repository

Reviewer 4: ```suggestion
- Clone the Azure Arc Jumpstart repository
```

    ```shell
    git clone https://github.com/microsoft/azure_arc.git
    ```

> * [Install or update Azure CLI to version 2.25.0 and above](https://docs.microsoft.com/en-us/cli/azure/install-azure-cli?view=azure-cli-latest). Use the below command to check your current installed version.

Reviewer 4: ```suggestion
- [Install or update Azure CLI to version 2.25.0 and above](https://docs.microsoft.com/en-us/cli/azure/install-azure-cli?view=azure-cli-latest). Use the below command to check your current installed version.
```

  ```shell
  az --version
  ```

> * Create Azure service principal (SP)

Reviewer 4: ```suggestion
- Create Azure service principal (SP)
```

    ```shell
    az login
    az ad sp create-for-rbac -n "<Unique SP Name>" --role contributor
    ```

    For example:

    ```shell
    az ad sp create-for-rbac -n "http://AzureArcK8s" --role contributor
    ```

    Output should look like this:

    ```json
    {
    "appId": "XXXXXXXXXXXXXXXXXXXXXXXXXXXX",
    "displayName": "AzureArcK8s",
    "name": "http://AzureArcK8s",
    "password": "XXXXXXXXXXXXXXXXXXXXXXXXXXXX",
    "tenant": "XXXXXXXXXXXXXXXXXXXXXXXXXXXX"
    }
    ```

    > **Note: The Jumpstart scenarios are designed with as much ease of use in-mind and adhering to security-related best practices whenever possible. It is optional but highly recommended to scope the service principal to a specific [Azure subscription and resource group](https://docs.microsoft.com/en-us/cli/azure/ad/sp?view=azure-cli-latest) as well considering using a [less privileged service principal account](https://docs.microsoft.com/en-us/azure/role-based-access-control/best-practices)**

## Automation Flow

> For you to get familiar with the automation and deployment flow, below is an explanation.
> 
> * User has deployed Kubernetes cluster using Cluster API and has it connected as Azure Arc-enabled Kubernetes cluster.
> 
> * User is editing the environment variables in the Shell script file (1-time edit) which then be used throughout the extension deployment.
> 
> * User is running the shell script. The script will use the extension management feature of Azure Arc to deploy the Azure Key Vault Secrets Provider extension on the Azure Arc connected cluster.
> 
> * The script will also deploy the Key Vault, sample secret to Azure Subscription, and a sample app to Azure Arc-enabled Kubernetes cluster
> 
> * User is veryfing the cluster and make sure the extension is deployed.
> 
> * User is checking the secret value mounted in pod.

Reviewer 4: ```suggestion
For you to get familiar with the automation and deployment flow, below is an explanation.

- User has deployed the Kubernetes cluster using Cluster API and has it connected as Azure Arc-enabled Kubernetes cluster.

- User is editing the environment variables in the Shell script file (1-time edit) which then be used throughout the extension deployment.

- User is running the shell script. The script will use the extension management feature of Azure Arc to deploy the Azure Key Vault Secrets Provider extension on the Azure Arc connected cluster.

- The script will also deploy the Key Vault, sample secret to Azure Subscription, and a sample app to Azure Arc-enabled Kubernetes cluster

- User is verifying the cluster and making sure the extension is deployed.

- User is checking the secret value mounted in a Kubernetes pod.
```

> ## Create Azure Key Vault extensions instance
> 
> To create a new extension Instance, we will use the _k8s-extension create_ command while passing in values for the mandatory parameters. This scenario provides you with the automation to deploy the Azure Key Vault extension on your Azure Arc-enabled Kubernetes cluster.
> 
> > **Note: Before installing the Key Vault extension, make sure that the kubectl context is pointing to your Azure Arc-enabled Kubernetes cluster. To do that, you can refer to the [official Kubernetes documentation](https://kubernetes.io/docs/tasks/access-application-cluster/configure-access-multiple-clusters/) to find the options to change the kubecontext to different Kubernetes clusters.**

Reviewer 4: ```suggestion
## Create Azure Key Vault extensions instance

To create a new extension Instance, we will use the _k8s-extension create_ command while passing in values for the mandatory parameters. This scenario provides you with the automation to deploy the Azure Key Vault extension on your Azure Arc-enabled Kubernetes cluster.

> **Note: Before installing the Key Vault extension, make sure that the _kubectl_ context is pointing to your Azure Arc-enabled Kubernetes cluster. To do that, you can refer to the [official Kubernetes documentation](https://kubernetes.io/docs/tasks/access-application-cluster/configure-access-multiple-clusters/) to find the options to change the kubecontext to different Kubernetes clusters.**
```

![Screenshot showing current kubectl context pointing to CAPI cluster](./01.png)

> * In the screenshot below, notice how currently there are no extensions installed yet in your Arc-enabled Kubernetes cluster.

Reviewer 4: ```suggestion
- In the screenshot below, notice how currently there are no extensions installed yet in your Arc-enabled Kubernetes cluster.
```

    ![Screenshot showing Azure Portal with Azure Arc-enabled Kubernetes resource extensions](./02.png)

> * Edit the environment variables in the script to match your environment parameters followed by running the ```. ./azure_keyvault_k8s_extension.sh``` command.

Reviewer 4: ```suggestion
- Edit the environment variables in the script to match your environment parameters followed by running the ```. ./azure_keyvault_k8s_extension.sh``` command.
```

    ![Screenshot parameter examples](./03.png)

    > **Note: The extra dot is due to the shell script has an *export* function and needs to have the vars exported in the same shell session as the rest of the commands.**

>    The script will:
> 
>   * Login to your Azure subscription using the SPN credentials
>   * Add or Update your local _connectedk8s_ and _k8s-extension_ Azure CLI extensions
>   * Create Resource Group, Key Vault and sample secret
>   * Create Azure Key Vault k8s extension instance
>   * Create Kubernetes Secret for supporting SPN Authentication
>   * Create Secret Provider Class to read the secrets
>   * Deploy the app

Reviewer 4: ```suggestion
   The script will:

  * Login to your Azure subscription using the SPN credentials
  * Add or Update your local _connectedk8s_ and _k8s-extension_ Azure CLI extensions
  * Create Resource Group, Key Vault and a sample secret
  * Create Azure Key Vault k8s extension instance
  * Create Kubernetes Secret for supporting SPN Authentication
  * Create Secret Provider Class to read the secrets
  * Deploy the app
```

> * You can now see that Azure Key Vault extensions are now enabled in the extension tab section of the Azure Arc-enabled Kubernetes cluster resource in Azure.
> 
>     ![Screenshot extension deployment extensions tab](./04.png)
> 
> * You can also verify the deployment by running the below _kubectl_ commands and see the deployed artifacts in the _kube-system_ namespace.

Reviewer 4: ```suggestion
- You can now see that Azure Key Vault extensions are now enabled in the extension tab section of the Azure Arc-enabled Kubernetes cluster resource in Azure.

    ![Screenshot extension deployment extensions tab](./04.png)

- You can also verify the deployment by running the below _kubectl_ commands and seeing the deployed artifacts in the _kube-system_ namespace.
```

    ```shell
    kubectl get pods,secret -n hello-arc
    kubectl get secretproviderclass -n hello-arc

    kubectl get customresourcedefinitions \
    secretproviderclasses.secrets-store.csi.x-k8s.io \
    secretproviderclasspodstatuses.secrets-store.csi.x-k8s.io
    ```

    ![Screenshot extension and app deployment](./05.png)

    ![Screenshot extension and app deployment](./06.png)

## Validate the Secrets

> * To verify that Azure Key Vault Secret Provider is working properly and fetching the secrets, run the below command to show secrets held in secrets-store:

Reviewer 4: ```suggestion
- To verify that Azure Key Vault Secret Provider is working properly and fetching the secrets, run the below command to show secrets held in secrets-store:
```

    ```shell
    kubectl -n hello-arc exec busybox-secrets-sync -- ls /mnt/secrets-store/
    ```

    ![Screenshot Kubernetes App Pod secret store](./07.png)

> * Run the below command to print a test secret held in secrets-store:

Reviewer 4: ```suggestion
- Run the below command to print a test secret held in secrets-store:
```

    ```shell
    kubectl -n hello-arc exec busybox-secrets-sync -- sh -c 'echo $SECRET_USERNAME'
    ```

    ![Screenshot Kubernetes App Pod secret](./08.png)

## Delete extension instance

> * The following command only deletes the extension instance.
> 
>     ```shell
>     az k8s-extension delete --name <extension-name> --cluster-type connectedClusters --cluster-name <cluster-name> --resource-group <resource-group>
>     ```
> 
> * You can also delete the extension from the Azure Portal under the extensions section of Azure Arc-enabled Kubernetes cluster resource.

Reviewer 4: ```suggestion
- The following command only deletes the extension instance.

    ```shell
    az k8s-extension delete --name <extension-name> --cluster-type connectedClusters --cluster-name <cluster-name> --resource-group <resource-group>
    ```

- You can also delete the extension from the Azure Portal under the extensions section of the Azure Arc-enabled Kubernetes cluster resource.
```
